bee-message/src/parents.rs

```diff
@@ -0,0 +1,188 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! The parents module defines the core data type for storing the messages directly approved by a message.
+
+use crate::{MessageId, MessagePackError, MessageUnpackError, ValidationError, MESSAGE_ID_LENGTH};
+
+use bee_ord::is_unique_sorted;
+use bee_packable::{PackError, Packable, Packer, UnpackError, Unpacker};
+
+use bitvec::prelude::*;
+
+use core::ops::{Deref, RangeInclusive};
+
+use alloc::{vec, vec::Vec};
+
+/// The range representing the valid number of parents.
+pub const MESSAGE_PARENTS_RANGE: RangeInclusive<usize> = 1..=8;
+
+/// Minimum number of strong parents for a valid message.
+pub const MESSAGE_MIN_STRONG_PARENTS: usize = 1;
+
+/// An individual message parent, which can be categorized as "strong" or "weak".
+#[derive(Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Serialize, serde::Deserialize),
+    serde(tag = "type", content = "data")
+)]
+pub enum Parent {
+    /// Message parents in which the past cone is "Liked".
+    Strong(MessageId),
+    /// Message parents in which the past cone is "Disliked", but the parents themselves are "Liked".
+    Weak(MessageId),
+}
+
+impl Parent {
+    /// Returns the `MessageId` of this `Parent`.
+    pub fn id(&self) -> &MessageId {
+        match self {
+            Self::Strong(id) => id,
+            Self::Weak(id) => id,
+        }
+    }
+}
+
+/// A [`Message`]'s `Parents` are the [`MessageId`]s of the messages it directly approves.
+///
+/// Parents must be:
+/// * in the `MESSAGE_PARENTS_RANGE` range;
+/// * lexicographically sorted;
+/// * unique;
+#[derive(Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct Parents {
+    inner: Vec<Parent>,
+}
+
+impl Deref for Parents {
+    type Target = [Parent];
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner.as_slice()
+    }
+}
+
+impl Parents {
+    /// Creates a new `Parents` instance from a given collection.
+    pub fn new(inner: Vec<Parent>) -> Result<Self, ValidationError> {
+        validate_parents_count(inner.len())?;
+        validate_parents_unique_sorted(&inner)?;
+
+        let strong_count = inner.iter().fold(0usize, |acc, parent| match parent {
+            Parent::Strong(_) => acc + 1,
+            _ => acc,
+        });
+
+        validate_strong_parents_count(strong_count)?;
+
+        Ok(Self { inner })
+    }
+
+    /// Returns the number of message parents.
+    pub fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    /// Returns true if the collection is empty.
+    pub fn is_empty(&self) -> bool {
```

Comment:
When created, Parents can't be empty so I'm not sure this method makes sense. It's probably been added to please clippy but I'd rather add an allow instead.

Reply:
Yep, makes sense to me.